component/pyroscope/ebpf/ebpf_linux.go

```diff
@@ -0,0 +1,266 @@
+//go:build linux
+
+package ebpf
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/discovery"
+	"github.com/grafana/agent/component/pyroscope"
+	ebpfspy "github.com/grafana/phlare/ebpf"
+	"github.com/grafana/phlare/ebpf/pprof"
+	"github.com/grafana/phlare/ebpf/sd"
+	"github.com/grafana/phlare/ebpf/symtab"
+	"github.com/oklog/run"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "pyroscope.ebpf",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			arguments := args.(Arguments)
+
+			targetFinder, err := sd.NewTargetFinder(os.DirFS("/"), opts.Logger, targetsOptionFromArgs(arguments))
+			if err != nil {
+				return nil, fmt.Errorf("ebpf target finder create: %w", err)
+			}
+			ms := newMetrics(opts.Registerer)
+
+			session, err := ebpfspy.NewSession(
+				opts.Logger,
+				targetFinder,
+				convertSessionOptions(arguments, ms),
+			)
+			if err != nil {
+				return nil, fmt.Errorf("ebpf session create: %w", err)
+			}
+
+			return New(opts, arguments, session, targetFinder, ms)
+		},
```

Comment:
Normally, the Build function just calls New, only passing the options and arugments to it. Does this extra logic need to be here or can it be moved to New?

Reply:
`session` and it's dependencies were created here to allow passing the `session`  as an argument to `New` function to be able to swap it in unit test. 

 I've moved and hardcoded the dependencies to the `New` function, now it is not  testable, so I've created another one in test `newTestComponent`